dev-tools/repl-utils/src/lib.rs

```diff
@@ -0,0 +1,157 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! helpers for running a REPL with reedline
+
+use anyhow::Context;
+use anyhow::anyhow;
+use anyhow::bail;
+use camino::Utf8Path;
+use clap::Parser;
+use reedline::Reedline;
+use reedline::Signal;
+use std::fs::File;
+use std::io::BufRead;
+use std::io::BufReader;
+
+/// Runs the same kind of REPL as `run_repl_on_stdin()`, but reads commands from
+/// a file
+///
+/// Commands are printed to stdout before being executed.
+///
+/// This is useful for expectorate tests.  You can define a file of input
+/// commands and check it against known output.  The output has the commands in
+/// it so that it's easier to verify.
+pub fn run_repl_from_file<C: Parser>(
+    input_file: &Utf8Path,
+    run_one: &mut dyn FnMut(C) -> anyhow::Result<Option<String>>,
```

Comment:
Is there an advantage to dynamic dispatch here, and in `run_repl_on_stdin`/`process_entry` below? Typically there would be just one implementation with monomorphization, I think?

Reply:
It's extra compile time + monomorphization for each caller vs. cost of dynamic dispatch dereference 🤷  I usually go with dynamic dispatch for closures like this because it's usually simpler and more concise.

Reply:
Totally hear you on compile time -- my point just was that since there's going to be just one caller, it's probably not going to matter. But this is a super duper nit so I'm fine deferring to you :)